### PR TITLE
[rclcpp] add WaitSet class and modify entities to work without executor

### PR DESCRIPTION
**Pull request by wjwwood**
This closes https://github.com/ros2/rclcpp/issues/520 or at the very least lays the ground work for closing it. This pull request adds a few new `rclcpp::*WaitSet` classes which can be used with all the "waitable" entities to wait on sets of them at the same time, using `rcl_wait_set_t` under the hood.

This pull request does these things:

- add guard condition class
  - replaces use of rcl_guard_condition_t directly
  - this pull request will probably not update all instances just yet, but examples can use it
- add message info class
  - replaces use of rmw_message_info_t, and updated all cases where it was being used before
- add wait set classes
  - main wait set class has synchronization and storage policies as template arguments
    - Policy based design, see: https://en.wikipedia.org/wiki/Modern_C%2B%2B_Design#Policy-based_design
  - add static and dynamic storage policy options
  - add thread-safe and sequential (not thread-safe) synchronization policy options
- update `Subscription` to have `take`, `take_type_erased`, and `take_serialized`
  - update `Executor` to use these rather than call rcl directly
  - user can now use these to take data without an executor
- update `Client` to have `take_response`
  - update `Executor` to use this instead
- update `Service` to have `take_request` and `send_response`
  - update `Executor` to use this instead

One thing mentioned in https://github.com/ros2/rclcpp/issues/520, which may not be in the first pass, is the ability to mix use of the executor and waitset together (like some subscriptions from a node are in an executor while others are not and instead use a waitset directly). This will be part of my other pull request to address https://github.com/ros2/rclcpp/issues/519.

What this pull request will _not_ do:

- update executor to use new wait set classes
  - I will do that in a follow up pr which also incorporates the use of time stamps from rcl
- change executors so more than one can be used with a node at a time

Things I would like to do, but will most likely have to be follow up work:

- add "algorithms" (utility functions) for working with `rclcpp::WaitSetResult`
  - in the spirit of `#include <algorithm>`, like `rclcpp::find_next_ready_subscription` kind of stuff
- support an allocator argument for the wait set
  - I spent a lot of time trying to integrate it, but it's complicated to support, but if we fell back to only supporting the polymorphic allocator it would be easier, but since we don't have c++17 and we haven't back ported the polymorphic allocator API to c++14 I couldn't easily do that yet.

----

TODO:

- [x] update to support client and service
- [x] finish thread-safe policy
- [x] add check to ensure entities are only added to one wait set at a time

I'm going move this out of draft mode so reviews can start while I finish the last few sets of changes.

Also, test coverage is missing in a few places, but I do have this example to exercise the main use cases:

https://github.com/ros2/examples/pull/262

I would appreciate help with adding more tests, but they are also something that I can add after the API freeze, so I'm de-prioritizing them now, as much as I hate to do that.

### CODE REVIEW

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/singlethreadedexecutor-creates-a-high-cpu-overhead-in-ros-2/10077/26

**Comment by alsora** [comment]
Hi, I would like to share results from some work that @mauropasse and I have been doing recently on "refactoring the executors"

This mainly started when we noticed spurious awakes when using the static executor in combination with intra-process communication https://github.com/nobleo/static_executor/issues/2.

For our tests, we used the latest ROS 2 master (i.e. right after the static executor got merged) and we run our 10 nodes 13 publisher single process application [see here the topology](https://github.com/irobot-ros/ros2-performance/blob/master/performances/benchmark/topology/cedar.json).
Running this application on a single core platform (RaspberryPi 1) resulted in **27% CPU usage**.

We did the following experiments to improve the results:
 1. Instead of adding the intra-process subscription waitable to the executor, we created a separate thread for each of these subscription that would act as its personal executor.
See [changes here](https://github.com/alsora/rclcpp/commit/ea3f97c2dded2fe4d33c36cae28efa4e11d34b1e)
This resulted in **19% CPU usage**.

 2. We updated the previous experiment by substituting the `rcl_wait` and `rcl_guard_condition_t` with `std::condition_variable`.
See [changes here](https://github.com/alsora/rclcpp/commit/609dd6e715628243d15c4e15ce11a84bbbfdee40).
This resulted in **15% CPU usage**.

3. Then, we applied the same principle also to timers, in the simplest possible way.
Instead of adding timers to the executor, we created for each publisher a separate thread where the publish function is periodically called using `std::chrono` instead of `rclcpp::Timer`.
This resulted in **10% CPU usage**.

These results show us the very high overhead that going through all the ROS 2 layers (rclcpp, rcl, rmw, dds) causes.

Because of this, we are looking at creating a so called "rclcpp executor", i.e. an executor that works entirely in the rclcpp layer and will do work for timers and intra-process subscriptions.
This executor should be templated to accept different syncrhonization primitives depending on the OS where you use it.
I think that the work that you are doing can definitely help in making the use of such executor very simple from the user perspective.

At the same time, we are also looking at how to reduce the overhead when for example publishing or receiving messages from a separate machine.

**Comment by wjwwood** [comment]
Thanks for your feedback @alsora! Hopefully the work I'm doing here will enable more improvements as you said. I'll try to comment briefly on your comments in detail.

> Instead of adding the intra-process subscription waitable to the executor, we created a separate thread for each of these subscription that would act as its personal executor.

That's really a decision about application layout, that is, in my opinion, very application dependent.

The wait set class I'm adding in this pull request and the change to allow multiple executors will allow users to decide how to map their waitable entities to threads how ever they like. In this pull request, you can add subscriptions in part or whole, so it should be possible to use one wait set for inter-process communication and another for intra-process communication even with a single subscription, but I haven't figured out how to scale that up to the executor just yet.

As for whether or not this should be the default, I can't answer that yet, at least not based on your data presented so far. I would have expected the latency to be reduced, but not the CPU and memory overhead, as more threads usually means more context switching (costs CPU) and more memory usage.

However, I'll keep this in mind and I'll depend on you guys and other in the community to test out the new design with respect to this point.

> We updated the previous experiment by substituting the `rcl_wait` and `rcl_guard_condition_t` with `std::condition_variable`.

I haven't had time to look at your changes, but I assume that means you cannot mix waiting on things like inter-process communication and QoS events from the same thread, since you need to wait on the condition variable rather than `rcl_wait`. I'd have to investigate why the guard condition used in the current intra-process communication is so expensive first.

Again, I'll keep it in mind and we'll have to check this again in the new architecture.

> Then, we applied the same principle also to timers, in the simplest possible way.
Instead of adding timers to the executor, we created for each publisher a separate thread where the publish function is periodically called using `std::chrono` instead of `rclcpp::Timer`.

Ok, this is good to know, but again it's not immediately obvious why this is the case. It seems to me that many threads would not scale well, but perhaps on the order or ~13 or so threads it doesn't matter.

Also, again this is something that is possible for the application to choose to do. We should (will) strive to make the ROS timers as efficient as possible, but there's nothing saying you must used them all the time. They're mostly a convenience when you don't want to create a dedicated thread to wait in, as you have done here.

> These results show us the very high overhead that going through all the ROS 2 layers (rclcpp, rcl, rmw, dds) causes.

I can certainly believe that is the case, but I would want to do some more investigation personally before coming to that conclusion. Each of the steps above changed a lot about the code paths being traversed. It's hard to say that it is only (or at least dominated by) the cost of traversing the API layers. Perhaps you're avoiding something inefficient in the C++ code instead?

I'll keep all of this in mind and hopefully get to trying your micro benchmarks towards the end of this process. I think that my refactors will make it easier to do some of the things you guys are trying and to enable other patterns of use, e.g. multiple executors or no executors (using the wait set directly). It will not fix all the issues you guys have pointed out. So again, I'll keep all of this in mind, but I expect we will need to do a few more passes over the code after these refactors land.

> Because of this, we are looking at creating a so called "rclcpp executor", i.e. an executor that works entirely in the rclcpp layer and will do work for timers and intra-process subscriptions.

I'd have to see what that looks like, but honestly it just sounds like something we could integrate into the a single executor, something between the single and multi threaded executors, which has one thread for things that use `rcl_wait` and another for things that don't need to use `rcl_wait`. I'd instead prefer to fix the issues with `rcl_wait`, but assuming that's not possible, this could be a good thing to have.

> This executor should be templated to accept different syncrhonization primitives depending on the OS where you use it.

The new wait set class is already using a policy based design that would allow for this.

**Comment by alsora** [comment]
We are currently investigating how conditions and timers are used in the lower layers of the stack, to understand if there is something that can be improved or if all the overhead is simply from having to traverse many layers.

The results that I shared with you use a separate thread for each entity and I would be surprised if they scaled well, nevertheless they show that there is a bottleneck somewhere in the current approach.

The `rclcpp executor` that we are trying to create will use a single thread (leaving its creation to the application as it's currently the case) and will have the purpose to be an easy to test proof of concept (what is done in the links above is just a bunch of quick hacks to see if the idea was meaningful or not).
Having a "2 threads executor", where 1 thread manages the RCLCPP entities (timers and intra-process communication) and the other one the RMW ones (pub/sub with other processes) is definitely a possible solution.

I will keep you posted when we have something you can test or more details about where the bottleneck exactly is.

File: rclcpp/include/rclcpp/waitable.hpp [hunk 1]
```diff
@@ -94,7 +94,6 @@ class Waitable
   size_t
   get_number_of_ready_guard_conditions();
 
-  // TODO(jacobperron): smart pointer?
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/waitable.hpp:97]: I looked into this, and in my opinion it should not be a smart pointer. @jacobperron unless you disagree I'll leave this removed.

File: rclcpp/src/rclcpp/executor.cpp [hunk 2]
```diff
@@ -305,54 +305,56 @@ Executor::execute_any_executable(AnyExecutable & any_exec)
...
 
+static
+void
+take_and_do_error_handling(
+  const char * action_description,
+  const char * topic_or_service_name,
+  std::function<bool()> take_action,
+  std::function<void()> handle_action)
+{
+  bool taken;
```
Comment (jacobperron) [inline rclcpp/src/rclcpp/executor.cpp:316]:
```suggestion
  bool taken = false;
```

File: rclcpp/include/rclcpp/wait_set_policies/detail/storage_policy_common.hpp [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+  get_raw_pointer_from_smart_pointer(const std::shared_ptr<EntityT> & shared_pointer)
+  {
+    return {nullptr, shared_pointer.get()};
+  }
+
+  template<class EntityT>
+  std::pair<std::shared_ptr<EntityT>, EntityT *>
+  get_raw_pointer_from_smart_pointer(const std::weak_ptr<EntityT> & weak_pointer)
+  {
+    auto shared_pointer = weak_pointer.lock();
```
[the remark this answers is not in the document]
Comment (wjwwood) [reply rclcpp/include/rclcpp/wait_set_policies/detail/storage_policy_common.hpp:133]: I don't think so, if it is expired, then this will return nullptr and the calling code can deal with that case.

File: rclcpp/include/rclcpp/wait_set_policies/dynamic_storage.hpp [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        return weak_ptr.expired();
+      };
+    // remove guard conditions which have been deleted
+    guard_conditions_.erase(std::remove_if(guard_conditions_.begin(), guard_conditions_.end(), p));
+  }
+
+  void
+  storage_acquire_ownerships()
+  {
+    if (++ownership_reference_counter_ > 1) {
```
[the remark this answers is not in the document]
Comment (wjwwood) [reply rclcpp/include/rclcpp/wait_set_policies/dynamic_storage.hpp:390]: If `ownership_reference_counter_` is `0`, then after `++` it will be `1`, which is `== 1`, but not `> 1`. I think it makes sense, but please check my logic.

File: rclcpp/include/rclcpp/wait_set_policies/sequential_synchronization.hpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    // Setup looping predicate.
+    auto start = std::chrono::steady_clock::now();
+    std::function<bool ()> should_loop = this->create_loop_predicate(time_to_wait_ns, start);
+
+    // Wait until exit condition is met.
+    do {
+      // Rebuild the wait set.
+      // This will resize the wait set if needed, due to e.g. adding or removing
+      // entities since the laster wait, but this should never occur in static
```
[the remark this answers is not in the document]
Comment (wjwwood) [reply rclcpp/include/rclcpp/wait_set_policies/sequential_synchronization.hpp]: This is sleep deprivation right here, haha. 42c2ce6

File: rclcpp/include/rclcpp/wait_set_template.hpp [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+   * This function will return when either one of the entities within this wait
+   * set is ready, or a period of time has passed, which ever is first.
+   * The term "ready" means different things for different entities, but
+   * generally it means some condition is met asynchronously for which this
+   * function waits.
+   *
+   * This function can either wait for a period of time, do no waiting
+   * (non-blocking), or wait indefinitely, all based on the value of the
+   * time_to_wait parameter.
+   * Waiting is always measured against the std::chrono::stead_clock.
```
[the remark this answers is not in the document]
Comment (wjwwood) [reply rclcpp/include/rclcpp/wait_set_template.hpp]: 240af99

File: rclcpp/src/rclcpp/executor.cpp [hunk 7]
```diff
@@ -306,82 +306,114 @@ Executor::execute_any_executable(AnyExecutable & any_exec)
...
+  message_info.get_rmw_message_info().from_intra_process = false;
+
+  // Reduce code duplication by generalize error handling into lambda.
+  auto take_message_and_do_error_handling =
+    [&subscription](
+      const char * message_adjective,
+      std::function<bool ()> take_action,
+      std::function<void ()> handle_action)
+    {
+      bool taken;
```
[the remark this answers is not in the document]
Comment (wjwwood) [reply rclcpp/src/rclcpp/executor.cpp]: 5f0dcf7

File: rclcpp/include/rclcpp/client.hpp [hunk 8]
```diff
@@ -171,6 +192,25 @@ class Client : public ClientBase
...
+   *   which the middleware will copy the response being taken.
+   * \param[out] request_header_out The request header to be filled by the
+   *   middleware when taking, and which can be used to associte the response
+   *   to a specific request.
+   * \returns true if the response was taken, otherwise false.
+   * \throws rclcpp::exceptions::RCLError based exceptions if the underlying
+   *   rcl function fail.
+   */
+  bool
+  take_response(typename ServiceT::Response & response_out, rmw_request_id_t & request_header_out)
```
[the remark this answers is not in the document]
Comment (wjwwood) [reply rclcpp/include/rclcpp/client.hpp:209]: Yeah, safety for users that don't need to use it as the base class, like in the wait set demos I made.

**Comment by wjwwood** [comment]
Preliminary CI up to `rclcpp`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10046/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5642/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8176/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/9956/)

File: rclcpp/include/rclcpp/wait_set_policies/dynamic_storage.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacobperron) [reply rclcpp/include/rclcpp/wait_set_policies/dynamic_storage.hpp:395]: Makes sense, I must have had `>= 1` in my mind.

File: rclcpp/include/rclcpp/wait_set_template.hpp [hunk 9]
```diff
@@ -0,0 +1,745 @@
...
+   */
+  void
+  add_subscription(
+    std::shared_ptr<rclcpp::SubscriptionBase> subscription,
+    rclcpp::SubscriptionWaitSetMask mask = {})
+  {
+    if (nullptr == subscription) {
+      throw std::invalid_argument("subscription is nullptr");
+    }
+    // this method comes from the SynchronizationPolicy
```
Comment (jacobperron) [inline rclcpp/include/rclcpp/wait_set_template.hpp:137]: To be clear, there's nothing technically preventing me from defining this method in my StoragePolicy class, right?

E.g. I could just define all of my sync and storage policy methods in one class and pass a dummy class as the second template.

Do you think it's worth trying to enforce that users implement the appropriate methods by defining abstract interfaces classes `SynchronizationPolicy` and `StoragePolicy`? Then these interfaces would make it clear what is expected of them. I don't feel strongly about this, and we can leave it as-is. Just making sure I understand what's going on.

File: rclcpp/test/test_wait_set.cpp [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+  {
+    auto gc = std::make_shared<rclcpp::GuardCondition>();
+    rclcpp::WaitSet wait_set;
+    ASSERT_THROW(
+    {
+      wait_set.remove_guard_condition(gc);
+    }, std::runtime_error);
+  }
+
+  // remove after delete, checking weak ownership behavior
```
Comment (jacobperron) [inline rclcpp/test/test_wait_set.cpp:204]: I think you should uncomment this test case, or remove it if it's not something we can really test.

File: rclcpp/include/rclcpp/wait_set_template.hpp (hunk 9 above)
Comment (wjwwood) [reply rclcpp/include/rclcpp/wait_set_template.hpp:137]: Yes you could move the functions around in the various template arguments.

If we had new features I could use something like concepts to ensure the template arguments met a certain interface more easily. Without that, I would have to create a virtual class here, which I was explicitly trying to avoid so that there were no virtual function calls in some of these performance sensitive areas. That's one of the advantages of the policy based design. Otherwise there'd be no need to have template arguments, instead we could just use normal inheritance.

File: rclcpp/test/test_wait_set.cpp (hunk 10 above)
Comment (wjwwood) [reply rclcpp/test/test_wait_set.cpp:204]: I'll delete it then I guess, it is a case I thought would be a corner case but for an non-obvious reason wasn't actually a problem. I guess I was leaving it there so in the future if someone also thought that should be a test they might see this and know we've thought about it already, but I guess that's not likely.

Comment (wjwwood) [reply rclcpp/test/test_wait_set.cpp]: f0dcf15

**Comment by wjwwood** [comment]
New CI after fixing some Windows issues:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10073/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5661/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8191/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/9976/)

**Comment by wjwwood** [comment]
> Although, I can't tell if the test_rate failure on Windows is new.

I believe it is not, since it's timing related, and I don't think this pr should affect rates, just timers.

**Comment by wjwwood** [comment]
Rebase and CI (up to rclcpp and the examples package):

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10134/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5704/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8237/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10030/)

**Comment by wjwwood** [comment]
The CI looks good, I'll merge this pr and continue to improve the demos at https://github.com/ros2/examples/pull/262

**Comment by rotu** [comment]
This seems to have broken `test_rclcpp` as well https://ci.ros2.org/job/ci_linux/10146/console#console-section-28

**Comment by wjwwood** [comment]
Sorry for the disruption, I had several sets of changes staged and forgot about necessary changes in rosbag2 and system_tests for this pull request, so locally things built. I was trying to save time on CI in my first CI runs, since it's been busy, but I had intended to run full CI afterwards, but did not. So, it's my fault entirely.

I have open pull requests https://github.com/ros2/system_tests/pull/423 and https://github.com/ros2/rosbag2/pull/375 (though I closed it in favor of https://github.com/ros2/rosbag2/pull/373) to address the issue.

Full CI with just Fast-RTPS:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10161/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5723/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8259/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10049/)

If that is "green" (or yellow with unrelated or known issues), I'll merge those prs to get things going again on CI. And then I'll run full CI with all rmw implementations and address any feedback if I don't get reviews on the other prs in time.

**Comment by wjwwood** [comment]
https://github.com/ros2/ros1_bridge/pull/253 is also needed, but only for packaging jobs, see the CI linked there for that.

**Comment by wjwwood** [comment]
Ok, I triaged the CI results and I think it should be merged. I'll follow up on the `laser_geometry` timeout issue.

**Comment by alsora** [comment]
@wjwwood is there a plan to update the executors to use the new rclcpp waitset class?

**Comment by wjwwood** [comment]
Yes, I'll do it with the pr that closes this: https://github.com/ros2/rclcpp/issues/519

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2020-04-16/13709/1

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/tackling-of-flaky-tests/13686/2

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/official-java-client-library/13389/17

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/improving-ros-2-cpu-by-simplifying-the-ros-2-layers/13808/1

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2020-05-21/14247/1

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-foxy-fitzroy-released/14495/1

**Comment by dejanpan** [comment]
## WaitSet Presentation at RTWG
1. Video recording: https://drive.google.com/open?id=1I6caR6MrfQOUvylO4fFG_zHMW4J62u7k 
1. Example of use: https://github.com/ros2/examples/pull/262 for this implementation: https://github.com/ros2/rclcpp/pull/1047.
1. 1st reason for WaitSet: users can use things that you wait on, without an executor and decide what to do with these things
1. 2nd reason for WaitSet: to move the executor forward => simplify it and make it more efficient by avoiding building of WaitSets, ...
   1. wait function tends to be in the common loop => we needed to make it performant, memory allocations
1. 2 basic things that WaitSet does: waiting and taking the data (the latter is normally done in the Executor)
1. Power of WaitSet - user can explicitly decide in what order to do things and whether to wait again. Executor is the encapsultation of those decisions.
1. Callbacks are not yet optional but will be
1. C++ version of the guard condition => for interuption of the WaitSet
1. WaitSet Can be init-ed with the list of input arguments
1. You can add subscriptions or guard conditions after the creation
1. Wait method takes the timeout and returns the wait results (status condition (timeout, ready, empty), RAII (ensures wait-set is not mutated), 
1. Why do we need WaitSet in rclcpp 
   1. nicer to use
   1. manages the lifetime
   1. provides opportunity to explicitly do thread safety
1. How rcl::waitsets work
   1. create an array of pointers to waitables 
   1. things that are ready will point to thos objects, things that are not ready will be NULL
1. You can explicitly remove the guard conditions
1. WaitSet API will be integrated with the executor API
   1. TODO: was this done?
1. Several kinds of WaitSets:
    1. WaitSetTemplate => takes policies, provides interface, the details can be delegated to the policies (e.g. synchronization (e.g. sequential - no thread safety) and storage (dynamic storage for adding and removing stuff on the read time))
    1. StaticWaitSet - no removing or adding of guard conditions after the construction. This is as close to the Apex.AI implementation as possible. Allows setup and steady phases.
    1. ThreadSafeWaitSet - to be executed with the executor. 
        1. you can add/remove things, you can wait, they give you the result => comes with the overhead
1. Can you access the timestamp in the WaitSet => no, it is part of the message info. 
1. Implementation
    1. WaitSet template class: https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/wait_set_template.hpp
1. Why policy-based design (and not just to use inheritance)
    1. Because you need to use virtual functions (performance hit)
    1. Flexibility
1. Policies
    1. Sequential and ThreadSafe
    1. Implementation: https://github.com/ros2/rclcpp/tree/master/rclcpp/include/rclcpp/wait_set_policies
1. TODO: Integration of WaitSet with the Executor: https://github.com/ros2/rclcpp/issues/1287

**Comment by dejanpan** [comment]
@wjwwood I added the transcript above. I realize that we it would be great to convert above + https://github.com/ros2/rclcpp/pull/1047#issue-397007017 into a design document - what do you say?

One question that I do not understand. Why does the `WaitSet` need to be integrated with the Executor which supposedly will be done in #1287?  Is this because it is not implemented in the combination with the `PollingSubscription` as at Apex.AI internally?

cc @budrus

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros2-latency-using-different-node-setups/18335/12
